File: src/SynDiffix/Anonymizer.fs
module SynDiffix.Anonymizer

open System
open System.Security.Cryptography

type ContributionsState = { AidContributions: Dictionary<AidHash, float>; mutable UnaccountedFor: float }

type CountResult = { AnonymizedCount: int64; NoiseSD: float }

// ----------------------------------------------------------------
// Random & noise
// ----------------------------------------------------------------

// The noise seeds are hash values.
// From each seed we generate a single random value, with either a uniform or a normal distribution.
// Any decent hash function should produce values that are uniformly distributed over the output space.
// Hence, we only need to limit the seed to the requested interval to get a uniform random integer.
// To get a normal random float, we use the Box-Muller method on two uniformly distributed integers.

let private randomUniform (interval: Interval) (seed: Hash) =
  let randomUniform = uint32 ((seed >>> 32) ^^^ seed)
  // While using modulo to bound values produces biased output, we are using very small ranges
  // (typically less than 10), for which the bias is insignificant.
  let boundedRandomUniform = randomUniform % uint32 (interval.Upper - interval.Lower + 1)
  interval.Lower + int boundedRandomUniform

let private randomNormal stdDev (seed: Hash) =
  let u1 = float (uint32 seed) / float UInt32.MaxValue
  let u2 = float (uint32 (seed >>> 32)) / float UInt32.MaxValue
  let randomNormal = Math.Sqrt(-2.0 * log u1) * Math.Sin(2.0 * Math.PI * u2)
  stdDev * randomNormal

let private cryptoHashSaltedSeed salt (seed: Hash) : Hash =
  use sha256 = SHA256.Create()
  let seedBytes = BitConverter.GetBytes(seed)
  let hash = sha256.ComputeHash(Array.append salt seedBytes)
  BitConverter.ToUInt64(hash, 0)

let seedFromAidSet (aidSet: AidHash seq) = Seq.fold (^^^) 0UL aidSet

let private mixSeed (text: string) (seed: Hash) = text |> Hash.string |> ((^^^) seed)

let private generateNoise salt stepName stdDev noiseLayers =
  noiseLayers
  |> Seq.map (cryptoHashSaltedSeed salt >> mixSeed stepName >> randomNormal stdDev)
  |> Seq.reduce (+)

// ----------------------------------------------------------------
// Depth limiting
// ----------------------------------------------------------------

let noisyRowLimit salt (seed: Hash) nRows rowFraction =
  let rowLimitWithoutNoise = nRows / rowFraction

  // Select an integer between plus and minus 5% of `rowLimitWithoutNoise`.
  let noiseOffsetScale = rowLimitWithoutNoise / 20

  seed
  |> cryptoHashSaltedSeed salt
  |> mixSeed "precision_limit"
  |> randomUniform { Lower = -noiseOffsetScale; Upper = noiseOffsetScale }
  |> (+) rowLimitWithoutNoise

// ----------------------------------------------------------------
// AID processing
// ----------------------------------------------------------------

// Compacts flattening intervals to fit into the total count of contributors.
// Both intervals are reduced proportionally, with `topCount` taking priority.
// `None` is returned in case there's not enough AIDVs for a sensible flattening.
// `public` just to test the low-level algorithm
let private compactFlatteningIntervals outlierCount topCount totalCount =
  if totalCount < outlierCount.Lower + topCount.Lower then
    None
  else
    let totalAdjustment = outlierCount.Upper + topCount.Upper - totalCount

    let compactIntervals =
      if totalAdjustment <= 0 then
        outlierCount, topCount // no adjustment needed
      else
        // NOTE: at this point we know `0 < totalAdjustment <= outlierRange + topRange` (*)
        //       because `totalAdjustment = outlierCount.Upper + topCount.Upper - totalCount
        //                               <= outlierCount.Upper + topCount.Upper - outlierCount.Lower - topCount.Lower`
        let outlierRange = outlierCount.Upper - outlierCount.Lower
        let topRange = topCount.Upper - topCount.Lower
        // `topAdjustment` will be half of `totalAdjustment` rounded up, so it takes priority as it should
        let outlierAdjustment = totalAdjustment / 2
        let topAdjustment = totalAdjustment - outlierAdjustment

        // adjust, depending on how the adjustments "fit" in the ranges
        match outlierRange >= outlierAdjustment, topRange >= topAdjustment with
        | true, true ->
          // both ranges are compacted at same rate
          { outlierCount with Upper = outlierCount.Upper - outlierAdjustment },
          { topCount with Upper = topCount.Upper - topAdjustment }
        | false, true ->
          // `outlierCount` is compacted as much as possible by `outlierRange`, `topCount` takes the surplus adjustment
          { outlierCount with Upper = outlierCount.Lower },
          { topCount with Upper = topCount.Upper - totalAdjustment + outlierRange }
        | true, false ->
          // vice versa
          { outlierCount with Upper = outlierCount.Upper - totalAdjustment + topRange },
          { topCount with Upper = topCount.Lower }
        | false, false ->
          // Not possible. Otherwise `outlierRange + topRange < outlierAdjustment + topAdjustment = totalAdjustment` but we
          // knew the opposite was true in (*) above
          failwith "Impossible interval compacting."

    Some compactIntervals

type private AidCount = { FlattenedSum: float; Flattening: float; NoiseSD: float; Noise: float }

let inline private aidFlattening
  (anonContext: AnonymizationContext)
  (unaccountedFor: float)
  (aidContributions: (AidHash * ^Contribution) array)
  : AidCount option =
  let totalCount = aidContributions.Length
  let anonParams = anonContext.AnonymizationParams

  match compactFlatteningIntervals anonParams.OutlierCount anonParams.TopCount totalCount with
  | None -> None // not enough AIDVs for a sensible flattening
  | Some(outlierInterval, topInterval) ->
    let sortedAidContributions =
      aidContributions
      |> Array.sortByDescending (fun (aid, contribution) -> contribution, aid)

    let flatSeed =
      sortedAidContributions
      |> Seq.take (outlierInterval.Upper + topInterval.Upper)
      |> Seq.map fst
      |> seedFromAidSet
      |> cryptoHashSaltedSeed anonParams.Salt

    let outlierCount = flatSeed |> mixSeed "outlier" |> randomUniform outlierInterval
    let topCount = flatSeed |> mixSeed "top" |> randomUniform topInterval

    let topGroupSum =
      sortedAidContributions
      |> Seq.skip outlierCount
      |> Seq.take topCount
      |> Seq.sumBy snd

    let topGroupAverage = (float topGroupSum) / (float topCount)

    let flattening =
      sortedAidContributions
      |> Seq.take outlierCount
      |> Seq.map snd
      |> Seq.map (fun contribution -> (float contribution) - topGroupAverage)
      |> Seq.filter (fun contribution -> contribution > 0)
      |> Seq.sum

    let realSum = aidContributions |> Array.sumBy snd
    let flattenedUnaccountedFor = unaccountedFor - flattening |> max 0.
    let flattenedSum = float realSum - flattening
    let flattenedAvg = flattenedSum / float totalCount

    let noiseScale = max flattenedAvg (0.5 * topGroupAverage)
    let noiseSD = anonParams.LayerNoiseSD * noiseScale

    let noise =
      [ anonContext.BucketSeed; aidContributions |> Seq.map fst |> seedFromAidSet ]
      |> generateNoise anonParams.Salt "noise" noiseSD

    Some
      {
        FlattenedSum = flattenedSum + flattenedUnaccountedFor
        Flattening = flattening
        NoiseSD = noiseSD
        Noise = noise
      }

let private arrayFromDict (d: Dictionary<'a, 'b>) =
  d |> Seq.map (fun pair -> pair.Key, pair.Value) |> Seq.toArray

let private mapAidFlattening (anonContext: AnonymizationContext) perAidContributions =
  perAidContributions
  |> Array.map (fun aidState ->
    aidState.AidContributions
    |> arrayFromDict
    |> aidFlattening anonContext aidState.UnaccountedFor
  )

// Assumes that `byAidSum` is non-empty, meaning that there is at least one AID instance involved
let private anonymizedSum (byAidSum: AidCount seq) =
  let flattening =
    byAidSum
    // We might end up with multiple different flattened sums that have the same amount of flattening.
    // This could be the result of some AID values being null for one of the AIDs, while there were still
    // overall enough AIDs to produce a flattened sum.
    // In these case we want to use the largest flattened sum to minimize unnecessary flattening.
    |> Seq.maxBy (fun aggregate -> aggregate.Flattening, aggregate.FlattenedSum)

  let noise =
    byAidSum
    // For determinism, resolve draws using maximum absolute noise value.
    |> Seq.maxBy (fun aggregate -> aggregate.NoiseSD, Math.Abs(aggregate.Noise))

  (flattening.FlattenedSum + noise.Noise, noise.NoiseSD)

let private moneyRoundNoise noiseSD =
  if noiseSD = 0.0 then
    0.0
  else
    let roundingResolution = moneyRound (0.05 * noiseSD)

    (noiseSD / roundingResolution) |> ceil |> (*) roundingResolution

// ----------------------------------------------------------------
// Public API
// ----------------------------------------------------------------

/// Returns whether any of the AID value sets has a low count.
let isLowCount salt (lowCountParams: LowCountParams) (aidTrackers: (int64 * Hash) seq) =
  aidTrackers
  |> Seq.map (fun (count, seed) ->
    if count < lowCountParams.LowThreshold then
      true
    else
      let thresholdNoise = generateNoise salt "suppress" lowCountParams.LayerSD [ seed ]

      // `LowMeanGap` is the number of standard deviations between `LowThreshold` and desired mean
      let thresholdMean =
        lowCountParams.LowMeanGap * lowCountParams.LayerSD
        + float lowCountParams.LowThreshold

      let threshold = thresholdNoise + thresholdMean

      float count < threshold
  )
  |> Seq.reduce (||)

let countMultipleContributions (anonContext: AnonymizationContext) (contributions: ContributionsState array) =
  let byAid = mapAidFlattening anonContext contributions

  // If any of the AIDs had insufficient data to produce a sensible flattening
  // we have to abort anonymization.
  if byAid |> Array.exists ((=) None) then
    None
  else
    let (value, noiseSD) = byAid |> Array.choose id |> anonymizedSum

    Some
      {
        AnonymizedCount = value |> Math.roundAwayFromZero |> int64
        NoiseSD = moneyRoundNoise noiseSD
      }

let countSingleContributions (anonContext: AnonymizationContext) (count: int64) (seed: Hash) =
  [ anonContext.BucketSeed; seed ]
  |> generateNoise anonContext.AnonymizationParams.Salt "noise" anonContext.AnonymizationParams.LayerNoiseSD
  |> (+) (float count)
  |> Math.roundAwayFromZero
  |> int64

File: syndiffix/anonymizer.py
import hashlib
import math
import operator
import sys
from collections import Counter
from dataclasses import dataclass, field, replace
from functools import reduce
from typing import Iterator, cast

from .common import *


@dataclass
class PidContributions:
    value_counts: Counter[Hash] = field(default_factory=Counter)
    unaccounted_for: int = 0


@dataclass(frozen=True)
class CountResult:
    anonymized_count: int
    noise_sd: float


# ----------------------------------------------------------------
# Noise
# ----------------------------------------------------------------

# The noise seeds are hash values.
# From each seed we generate a single random value, with either a uniform or a normal distribution.
# Any decent hash function should produce values that are uniformly distributed over the output space.
# Hence, we only need to limit the seed to the requested interval to get a uniform random integer.
# To get a normal random float, we use the Box-Muller method on two uniformly distributed integers.


def _random_uniform(interval: FlatteningInterval, seed: Hash) -> int:
    # While using modulo to bound values produces biased output, we are using very small ranges
    # (typically less than 10), for which the bias is insignificant.
    return int(seed) % (interval.upper - interval.lower + 1) + interval.lower


def _random_normal(sd: float, seed: Hash) -> float:
    u1 = (int(seed) & 0x7FFFFFFF) / 0x7FFFFFFF
    u1 = max(u1, sys.float_info.epsilon)
    u2 = ((int(seed) >> 32) & 0x7FFFFFFF) / 0x7FFFFFFF
    normal = math.sqrt(-2.0 * math.log(u1)) * math.sin(2.0 * math.pi * u2)
    return sd * normal


def _crypto_hash_salted_seed(salt: bytes, seed: Hash) -> Hash:
    hash = hashlib.sha256(salt + seed.tobytes()).digest()
    return Hash(int.from_bytes(hash[:8], "little"))


def _hash_bytes(b: bytes) -> Hash:
    hash = hashlib.blake2b(b, digest_size=8).digest()
    return Hash(int.from_bytes(hash, "little"))


def _hash_string(s: str) -> Hash:
    return _hash_bytes(s.encode())


def _hash_int(i: int) -> Hash:
    return _hash_bytes(i.to_bytes(8, "little"))


def _mix_seed(step_name: str, seed: Hash) -> Hash:
    return _hash_string(step_name) ^ seed


def _generate_noise(salt: bytes, step_name: str, sd: float, noise_layers: tuple[Hash, ...]) -> float:
    noise = 0.0
    for layer_seed in noise_layers:
        noise += _random_normal(sd, _mix_seed(step_name, _crypto_hash_salted_seed(salt, layer_seed)))
    return noise


def _compact_flattening_intervals(
    outlier_count: FlatteningInterval, top_count: FlatteningInterval, total_count: int
) -> tuple[FlatteningInterval, FlatteningInterval] | None:
    if total_count < outlier_count.lower + top_count.lower:
        return None

    total_adjustment = outlier_count.upper + top_count.upper - total_count

    outlier_upper = outlier_count.upper
    top_upper = top_count.upper

    if total_adjustment > 0:
        # NOTE: At this point we know `0 < total_adjustment <= outlier_range + top_range` (*):
        #       `total_adjustment = outlier_count.upper + top_count.upper - total_count
        #                        <= outlier_count.upper + top_count.upper - outlier_count.lower - top_count.lower`.
        outlier_range = outlier_count.upper - outlier_count.lower
        top_range = top_count.upper - top_count.lower
        # `top_adjustment` will be half of `total_adjustment` rounded up, so it takes priority as it should.
        outlier_adjustment = total_adjustment // 2
        top_adjustment = total_adjustment - outlier_adjustment

        # Adjust, depending on how the adjustments "fit" in the ranges.
        match (outlier_range >= outlier_adjustment, top_range >= top_adjustment):
            case (True, True):
                # Both ranges are compacted at same rate.
                outlier_upper -= outlier_adjustment
                top_upper -= top_adjustment
            case (False, True):
                # `outlier_count` is compacted as much as possible by `outlier_range`,
                # `top_count` takes the surplus adjustment.
                outlier_upper = outlier_count.lower
                top_upper -= total_adjustment - outlier_range
            case (True, False):
                # Vice versa.
                outlier_upper -= total_adjustment - top_range
                top_upper = top_count.lower
            case (False, False):
                # Not possible. Otherwise:
                # `outlier_range + top_range < outlier_adjustment + top_adjustment = total_adjustment`,
                # but we knew the opposite was true in (*) above.
                raise RuntimeError("Impossible interval compacting.")

    return replace(outlier_count, upper=outlier_upper), replace(top_count, upper=top_upper)


@dataclass(frozen=True)
class _PidCount:
    flattened_count: float
    flattening: float
    noise_sd: float
    noise: float


def _flatten_contributions(pid_contributions: PidContributions, context: AnonymizationContext) -> _PidCount | None:
    total_count = len(pid_contributions.value_counts)
    anon_params = context.anonymization_params

    flattening_intervals = _compact_flattening_intervals(anon_params.outlier_count, anon_params.top_count, total_count)
    if flattening_intervals is None:
        return None  # Insufficient values for a sensible flattening.

    outlier_interval, top_interval = flattening_intervals

    # Sort contributions by amount and PID value (to ensure consistency in case of equal contributions).
    sorted_value_counts = sorted(
        pid_contributions.value_counts.items(),
        reverse=True,
        key=operator.itemgetter(1, 0),
    )

    flat_seed = seed_from_pid_set(pid for pid, _ in sorted_value_counts[: outlier_interval.upper + top_interval.upper])
    flat_seed = _crypto_hash_salted_seed(anon_params.salt, flat_seed)
    outlier_count = _random_uniform(outlier_interval, _mix_seed("outlier", flat_seed))
    top_count = _random_uniform(top_interval, _mix_seed("top", flat_seed))

    top_group_sum = sum(
        contribution for _, contribution in sorted_value_counts[outlier_count: (outlier_count + top_count)]
    )
    top_group_average = top_group_sum / top_count

    flattening = sum(
        max(contribution - top_group_average, 0) for _, contribution in sorted_value_counts[:outlier_count]
    )

    real_sum = pid_contributions.value_counts.total()
    flattened_unaccounted_for = max(pid_contributions.unaccounted_for - flattening, 0)
    flattened_sum = real_sum - flattening
    flattened_avg = flattened_sum / total_count

    noise_scale = max(flattened_avg, 0.5 * top_group_average)
    noise_sd = anon_params.layer_noise_sd * noise_scale

    pid_seed = seed_from_pid_set(pid_contributions.value_counts)
    noise = _generate_noise(anon_params.salt, "noise", noise_sd, (context.bucket_seed, pid_seed))

    return _PidCount(flattened_sum + flattened_unaccounted_for, flattening, noise_sd, noise)


def _anonymized_sum(pid_counts: Iterable[_PidCount]) -> tuple[float, float]:
    # We might end up with multiple different flattened counts that have the same amount of flattening.
    # This could be the result of some PID values being null for one of the PIDs, while there were still
    # overall enough PIDs to produce a flattened count.
    # In these cases, we want to use the largest flattened count to minimize unnecessary flattening.
    flattening = max(pid_counts, key=lambda count: (count.flattening, count.flattened_count))

    # For determinism, resolve draws using the maximum absolute noise value.
    noise = max(pid_counts, key=lambda count: (count.noise_sd, abs(count.noise)))

    return flattening.flattened_count + noise.noise, noise.noise_sd


MONEY_ROUND_MIN = 1e-10
MONEY_ROUND_DELTA = MONEY_ROUND_MIN / 100.0


# Works with `value` between 1.0 and 10.0.
def _money_round_internal(value: float) -> float:
    if 1.0 <= value < 1.5:
        return 1.0
    elif 1.5 <= value < 3.5:
        return 2.0
    elif 3.5 <= value < 7.5:
        return 5.0
    else:
        return 10.0


def _money_round(value: float) -> float:
    if 0.0 <= value < MONEY_ROUND_MIN:
        return 0.0

    tens = 10.0 ** math.floor(math.log10(value))
    return tens * _money_round_internal(value / tens)


def _money_round_noise(noise_sd: float) -> float:
    if noise_sd == 0.0:
        return 0.0

    rounding_resolution = _money_round(0.05 * noise_sd)
    return math.ceil(noise_sd / rounding_resolution) * rounding_resolution


# ----------------------------------------------------------------
# Public API
# ----------------------------------------------------------------


def hash_strings(strings: Iterator[str]) -> Hash:
    return seed_from_pid_set(_hash_string(string) for string in set(strings))


def hash_pid(pid: object) -> Hash:
    if not pid:
        return Hash(0)
    elif isinstance(pid, int):
        return _hash_int(cast(int, pid))
    elif isinstance(pid, str):
        return _hash_string(cast(str, pid))
    else:
        raise NotImplementedError("Unsupported PID type!")


def seed_from_pid_set(pid_set: Iterable[Hash]) -> Hash:
    return reduce(operator.xor, pid_set, Hash(0))


# Returns whether any of the PID value sets has a low count.
def is_low_count(salt: bytes, params: SuppressionParams, pid_trackers: list[tuple[int, Hash]]) -> bool:
    assert len(pid_trackers) > 0

    for count, seed in pid_trackers:
        if count < params.low_threshold:
            return True

        noise = _generate_noise(salt, "suppress", params.layer_sd, (seed,))

        # `low_mean_gap` is the number of standard deviations between `low_threshold` and desired mean.
        mean = params.low_mean_gap * params.layer_sd + params.low_threshold

        if count < noise + mean:
            return True

    return False


def count_multiple_contributions(
    context: AnonymizationContext, contributions_list: list[PidContributions]
) -> CountResult | None:
    assert len(contributions_list) > 0

    flattened_contributions = [_flatten_contributions(contributions, context) for contributions in contributions_list]

    # If any of the PIDs had insufficient data to produce a sensible flattening, we have to abort anonymization.
    if not all(flattened_contributions):
        return None

    value, noise_sd = _anonymized_sum(cast(list[_PidCount], flattened_contributions))

    return CountResult(round(value), _money_round_noise(noise_sd))


def count_single_contributions(context: AnonymizationContext, count: int, seed: Hash) -> int:
    params = context.anonymization_params
    noise = _generate_noise(params.salt, "noise", params.layer_noise_sd, (context.bucket_seed, seed))
    return round(count + noise)


def noisy_row_limit(salt: bytes, seed: Hash, row_count: int, row_fraction: int) -> int:
    real_row_limit = row_count // row_fraction

    # Select an integer between plus and minus 5% of `real_row_limit`.
    noise_range = real_row_limit // 20
    noise_seed = _mix_seed("precision_limit", _crypto_hash_salted_seed(salt, seed))
    noise = _random_uniform(FlatteningInterval(-noise_range, noise_range), noise_seed)

    return real_row_limit + noise
